### PR TITLE
Enable strictNullChecks for integration tests

### PR DESCRIPTION
**Pull request by tomduncalf**
## What, How & Why?

This change is required for flexible sync. Ideally we'd enable `strict: true` but this throws a lot more errors so just doing `strictNullChecks` for now.

I've tried to do runtime assertions where I thought it might make sense (e.g. there's a chance the thing might not exist, if something is badly broken, so throw) and otherwise use non-null assertions, but happy to change if people think we should take a different approach.

## ☑️ ToDos

* [x] 📝 Changelog entry
* [ ] 📝 `Compatibility` label is updated or copied from previous entry
* [ ] 🚦 Tests
* [ ] 📱 Check the React Native/other sample apps work if necessary
* [ ] 📝 Public documentation PR created or is not necessary
* [ ] 💥 `Breaking` label has been applied or is not necessary

*If this PR adds or changes public API's:*
* [ ] typescript definitions file is updated
* [ ] jsdoc files updated
* [ ] Chrome debug API is updated if API is available on React Native

### CODE REVIEW

File: integration-tests/tests/src/hooks/open-realm-before.ts [hunk 1]
```diff
@@ -46,6 +46,9 @@ export function openRealmHook(config: LocalConfiguration | SyncedConfiguration =
       } as Realm.Configuration;
       this.realm = new Realm(this.config);
       // Upload the schema, ensuring a valid connection
+      if (!this.realm.syncSession) {
+        throw new Error("No syncSession found on realm");
+      }
       await this.realm.syncSession.uploadAllLocalChanges();
```
Comment (kraenhansen) [inline integration-tests/tests/src/hooks/open-realm-before.ts:52]:
```suggestion
      if (this.realm.syncSession) {
        await this.realm.syncSession.uploadAllLocalChanges();
      }
```

Comment (tomduncalf) [reply integration-tests/tests/src/hooks/open-realm-before.ts:52]: My thinking was that if the `syncSession` does not exist here, something has gone badly wrong, as we've created a realm with a sync config, so we should throw

File: integration-tests/tests/src/hooks/open-realm-before.ts [hunk 2]
```diff
@@ -54,12 +57,16 @@ export function openRealmHook(config: LocalConfiguration | SyncedConfiguration =
 export function closeRealm(this: RealmContext): void {
   if (this.realm) {
     this.realm.close();
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore Ignore error as we ensure the realm is reopened
```
Comment (kraenhansen) [inline integration-tests/tests/src/hooks/open-realm-before.ts:61]: Perhaps `this` should be a `Partial<RealmContext>` instead?

File: integration-tests/tests/src/hooks/open-realm-before.ts (hunk 1 above)
Comment (kraenhansen) [reply integration-tests/tests/src/hooks/open-realm-before.ts:52]: Ahh sorry, I hadn't expanded the code snippet. I agree, we're in a branch where sync is expected.

File: integration-tests/tests/src/hooks/open-realm-before.ts (hunk 2 above)
Comment (tomduncalf) [reply integration-tests/tests/src/hooks/open-realm-before.ts:61]: Yeah that would work!

Comment (takameyer) [reply integration-tests/tests/src/hooks/open-realm-before.ts:61]: What was the ts error here?

Comment (tomduncalf) [reply integration-tests/tests/src/hooks/open-realm-before.ts:61]: `The operand of a 'delete' operator must be optional` which is a new one to me!

File: integration-tests/tests/src/tests/objects.ts [hunk 3]
```diff
@@ -33,7 +33,7 @@ describe("Realm objects", () => {
   describe("Interface & object literal", () => {
     it("can be created", () => {
       const realm = new Realm({ schema: [PersonSchema] });
-      let john: IPerson;
+      let john!: IPerson;
 
       realm.write(() => {
         john = realm.create<IPerson>(PersonSchema.name, {
```
Comment (kraenhansen) [inline integration-tests/tests/src/tests/objects.ts:39]: Actually, with https://github.com/realm/realm-js/pull/3937 merged, we might as well do:

```suggestion
      const john = realm.write(() => {
        return realm.create<IPerson>(PersonSchema.name, {
```

File: integration-tests/tests/src/tests/serialization.ts [hunk 4]
```diff
@@ -484,7 +484,8 @@ describe("JSON serialization", () => {
       afterEach(() => {
         if (realm) {
           realm.write(() => {
-            realm.deleteAll();
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            realm!.deleteAll();
```
Comment (kraenhansen) [inline integration-tests/tests/src/tests/serialization.ts:488]: I'm surprised this is needed ...

File: integration-tests/tests/src/hooks/open-realm-before.ts (hunk 2 above)
Comment (takameyer) [reply integration-tests/tests/src/hooks/open-realm-before.ts:61]: Ahhh...ok that makes sense.  Since `this` is typed with non-optionals.  So `Partial<RealmContext>` is appropriate

File: integration-tests/tests/src/utils/benchmark.ts [hunk 5]
```diff
@@ -35,7 +35,8 @@ export function itPerforms(title: string, fn: () => void, options?: Partial<Benc
       maximumFractionDigits: 0,
     });
     const sd = Number(result.sd.toFixed(2));
-    this.test.title += ` (${ops} ops/sec, ±${sd}%)`;
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    this.test!.title += ` (${ops} ops/sec, ±${sd}%)`;
```
Comment (kraenhansen) [inline integration-tests/tests/src/utils/benchmark.ts:39]: Does this fly? Using the `!` operator is really just hiding a potential issue.

```suggestion
    if (this.test) {
      this.test.title += ` (${ops} ops/sec, ±${sd}%)`;
    }
```

File: integration-tests/tests/src/tests/objects.ts (hunk 3 above)
Comment (takameyer) [reply integration-tests/tests/src/tests/objects.ts:39]: Why are we not using classes for integration tests?  (Question is out of scope for this PR, just curious)

File: integration-tests/tests/src/tests/serialization.ts (hunk 4 above)
Comment (tomduncalf) [reply integration-tests/tests/src/tests/serialization.ts:488]: Yeah I have been confused by this before. I believe the logic is that TS can't infer how the callback is called, so it could be being called async and `realm` has changed in the interim to being `null`

image

File: integration-tests/tests/src/utils/benchmark.ts (hunk 5 above)
Comment (tomduncalf) [reply integration-tests/tests/src/utils/benchmark.ts:39]: I was making the assumption that `test` would always be defined. Though if mocha type defs mark it as optional, I guess there maybe is some edge case! I think it might be appropriate to throw if `test` is undefined as it would mean something weird has happened?

File: integration-tests/tests/src/tests/serialization.ts (hunk 4 above)
Comment (kraenhansen) [reply integration-tests/tests/src/tests/serialization.ts:488]: But `realm` is captured when the lambda is created? It can't change value from that point on 🤔

File: integration-tests/tests/src/tests/objects.ts (hunk 3 above)
Comment (tomduncalf) [reply integration-tests/tests/src/tests/objects.ts:39]: Oh I had no idea #3937 was a thing, nice!

File: integration-tests/tests/src/tests/serialization.ts (hunk 4 above)
Comment (kraenhansen) [reply integration-tests/tests/src/tests/serialization.ts:488]: It seems like a TSC bug to me.

Comment (tomduncalf) [reply integration-tests/tests/src/tests/serialization.ts:488]: Hmm yeah... maybe the TS logic is a bit faulty, I think this is described in https://github.com/microsoft/TypeScript/issues/10982#issuecomment-247993896

Comment (kraenhansen) [reply integration-tests/tests/src/tests/serialization.ts:488]: It works as expected here: https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3nKEC2AXPCgK4QTwA+ABQN7wBmSSRARAEayvwC+A3ACgBASwbwAFDnwBKeLQHxFiVJkbN4AXmwhceAHRMkgpcvRZoedsCiaJsjQD45Ck0uTokEEHohIA5pI6+AbM0sZK-AI8AkA

File: integration-tests/tests/src/tests/objects.ts [hunk 6]
```diff
@@ -77,6 +77,7 @@ describe("Realm objects", () => {
       });
 
       const john = realm.objectForPrimaryKey<IPersonWithId>(PersonSchemaWithId.name, _id);
+      if (!john) throw new Error("Object not found");
```
Comment (takameyer) [inline integration-tests/tests/src/tests/objects.ts:80]: Can we `expect(john).to.not.be(undefined)`, or does that defeat the purpose of this PR?

File: integration-tests/tests/src/tests/serialization.ts (hunk 4 above)
Comment (takameyer) [reply integration-tests/tests/src/tests/serialization.ts:488]: I hope its not

Comment (tomduncalf) [reply integration-tests/tests/src/tests/serialization.ts:488]: That's weird...

Comment (kraenhansen) [reply integration-tests/tests/src/tests/serialization.ts:488]: Ahh. Perhaps https://github.com/realm/realm-js/pull/4193/files#diff-13ec3f38b0cf2f5f7842047684a331ad82a406f144a3e9ac84cad52718bd824eR491 is the offender :)

Comment (tomduncalf) [reply integration-tests/tests/src/tests/serialization.ts:488]: I think it's actually because `realm` is a `let` https://github.com/realm/realm-js/pull/4193/files#diff-13ec3f38b0cf2f5f7842047684a331ad82a406f144a3e9ac84cad52718bd824eR475

See https://www.typescriptlang.org/play?ssl=2&ssc=1&pln=2&pc=4#code/PTAEBMFMGMBsEMBOlTQPYDsDOAXUz5YBbALlAwFdZZQAfAAoG9QAzNNMgIgCMlPQAvgG4AULEh4CxMpWp0mrdl16J+A0AF5QzNh1A8+g0SICWLUAAopRAJTaRoR6ky5FaTfkiEiAOl2inZ2w8BCJucHgPCzsNAD57QMD0bDRxH1g0AHMrL2I-dhsAp2ERAREgA

**Review by takameyer — COMMENTED** [review]
I think the test should make assertions instead of throwing exceptions.  Maybe this warrants a  quick discussion.

File: integration-tests/tests/src/tests/serialization.ts (hunk 4 above)
Comment (kraenhansen) [reply integration-tests/tests/src/tests/serialization.ts:488]: Oh my ... I learned something today.

File: integration-tests/tests/src/tests/objects.ts (hunk 6 above)
Comment (tomduncalf) [reply integration-tests/tests/src/tests/objects.ts:80]: We could, but this wouldn't satisfy Typescript so we'd have to do non-null assertions or whatever. I thought that as its an edge case (like should never happen), throwing is the least lines of code required to both fail the test and keep TS happy, but the test error may be a little less clear if you did hit the edge case

**Review by takameyer — APPROVED** [review]

File: integration-tests/tests/src/tests/objects.ts (hunk 6 above)
Comment (takameyer) [reply integration-tests/tests/src/tests/objects.ts:80]: Ok...that makes sense.  We need to keep TS happy 🍼

**Review by kraenhansen — APPROVED** [review]
I would love for me or you (probably me) to do a follow-up PR making this `strict` instead of simply `strictNullChecks`.

**Comment by tomduncalf** [comment]
If you have the time, that would be awesome, but if not I am happy to look at it once my current bit of work is out of the way (probably end of the week). I'll create an issue to track

**Comment by tomduncalf** [comment]
https://github.com/realm/realm-js/issues/4200
